content/blog/2019-11-testthat-2-3-0/index.Rmarkdown

```diff
@@ -0,0 +1,146 @@
+---
+title: testthat 2.3.0
+author: Hadley Wickham
+date: '2019-11-04'
+slug: testthat-2-3-0
+categories:
+  - package
+tags:
+  - r-lib
+  - devtools
+  - testthat
+photo:
+  url: https://unsplash.com/photos/n_vD-7RxA3Q
+  author: Roman Kraft
+
+---
+
+```{r include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+We're pumped to announce that [testthat 2.3.0](http://testthat.r-lib.org) is now available on CRAN! 
+testthat makes it easy to turn your existing informal tests into formal automated tests that you can rerun quickly and easily. 
+testthat is the most popular unit-testing package for R, and is used by over 4,000 CRAN and Bioconductor packages. 
+You can learn more about unit testing at <https://r-pkgs.org/tests.html>. 
+
+(We didn't write a blog post about testthat 2.2.0 because it only introduced a single, experimental, new feature: `verify_output()`. 
+It has now matured to the point we think you should also try it out, so it's discussed below.)
+
+Install the latest version of testthat with:
+
+```{r, eval = FALSE}
+install.packages("testthat")
+```
+
+This release features two main improvements:
+
+* A general overhaul of condition and error handling.
+
+* A polished `verify_output()` that is ready for you to use to test your
+  print methods and error messages.
+
+For a complete list of changes, please see the [release notes](https://testthat.r-lib.org/news/index.html#testthat-2-3-0).  
+
+```{r setup}
+library(testthat)
+```
+
+## Errors
+
+The main improvements are mostly behind the scenes: we have overhauled the handling of errors and backtraces so that you should get more informative outputs when tests error unexpectedly or fail. 
+It's hard to authentically demonstrate this in an RMarkdown document, but if you have an error inside a test, like this:
+
+```{r, eval = FALSE}
+f <- function() g()
+g <- function() h()
+h <- function() stop("This is an error!")
+
+test_that("f() works as expected", {
+  expect_equal(f(), 10)
+})
+```
+
+You'll now get an informative backtrace that should allow you to quickly locate the source of the error:
+
+```
+test-error.R:6: error: f() works as expected
+This is an error!
+Backtrace:
+ 1. testthat::expect_equal(f(), 10) tests/testthat/test-error.R:6:2
+ 4. testthat:::f()
+ 5. testthat:::g() tests/testthat/test-catch.R:1:5
+ 6. testthat:::h() tests/testthat/test-catch.R:2:5
+```
+
+The previous version only showed the error message, which wasn't terribly useful!
```

Comment:
IIRC, the previous version would show a base-style backtrace in case of unexpected errors? Only failed error expectations would only show the unexpected error message.

Reply:
I did run the code with old testthat, and it only printed the error message.